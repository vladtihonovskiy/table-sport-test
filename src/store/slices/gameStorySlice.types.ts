export interface playersObjectTypes {
  name: string;
  played: number;
  win: number;
  draw: number;
  lost: number;
  totalPoints: number;
}

export interface GameObjectType {
  id: string;
  parentName: string;
  guestName: string;
  parentPoints: number;
  guestPoints: number;
  isPlayed: boolean;
}

export interface GameStoryState {
  listOPlayers: playersObjectTypes[];
  listOfGames: GameObjectType[];
}

export interface SaveGameScorePayload {
  id: string;
  parentPoints: number;
  guestPoints: number;
  parentName: string;
  guestName: string;
}
