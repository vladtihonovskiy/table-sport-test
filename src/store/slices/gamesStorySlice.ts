import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import {
  GameObjectType,
  GameStoryState,
  playersObjectTypes,
  SaveGameScorePayload,
} from "./gameStorySlice.types";

const newPlayerObjectCreator = (name: string) => ({
  name,
  played: 0,
  win: 0,
  draw: 0,
  lost: 0,
  totalPoints: 0,
});

const newGameObjectCreator = (parentName: string, guestName: string) => ({
  id: `${parentName} ${guestName}`,
  parentName,
  guestName,
  guestPoints: 0,
  parentPoints: 0,
  isPlayed: false,
});

const createNewGames = (
  listOfPlayers: playersObjectTypes[],
  newCommandName: string
) => {
  const newGames: GameObjectType[] = [];
  listOfPlayers.forEach((item) => {
    newGames.push(newGameObjectCreator(item.name, newCommandName));
  });
  return newGames;
};

const initialState: GameStoryState = {
  listOPlayers: [],
  listOfGames: [],
};

export const gameStorySlice = createSlice({
  name: "gameStory",
  initialState,
  reducers: {
    addNewPlayerName: (state, action: PayloadAction<string>) => {
      if (state.listOPlayers.length >= 1) {
        const newListOfGames = createNewGames(
          [...state.listOPlayers],
          action.payload
        );
        state.listOfGames.push(...newListOfGames);
      }
      state.listOPlayers.push(newPlayerObjectCreator(action.payload));
    },
    saveGameResult: (state, action: PayloadAction<SaveGameScorePayload>) => {
      const { id, parentPoints, guestPoints, parentName, guestName } =
        action.payload;
      const updateGameIndex = state.listOfGames.findIndex(
        (item) => item.id === id
      );
      state.listOfGames[updateGameIndex] = {
        ...state.listOfGames[updateGameIndex],
        parentPoints,
        guestPoints,
        isPlayed: true,
      };

      // const parentPlayerIndex = state.listOPlayers.findIndex(
      //   (item) => item.name === parentName
      // );
      // const guestPlayerIndex = state.listOPlayers.findIndex(
      //   (item) => item.name === guestName
      // );
    },
  },
});

// Action creators are generated for each case reducer function
export const { addNewPlayerName, saveGameResult } = gameStorySlice.actions;

export default gameStorySlice.reducer;
