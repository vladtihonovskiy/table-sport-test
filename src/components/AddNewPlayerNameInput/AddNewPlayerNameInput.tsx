import React, { useCallback, useState } from "react";
import { Button, Grid, TextField } from "@mui/material";
import { toast } from "react-toastify";
import { useAppDispatch } from "../../hooks/useAppDispatch";
import { addNewPlayerName } from "../../store/slices/gamesStorySlice";
import { useAppSelector } from "../../hooks/useAppSelector";

export const AddNewPlayerNameInput = () => {
  const dispatch = useAppDispatch();
  const { listOPlayers } = useAppSelector((state) => state.gameStory);

  const [textInputValue, setTextInputValue] = useState("");

  const onInputHandleChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setTextInputValue(e.target.value);
    },
    []
  );

  const onAddButtonHandleClick = useCallback(() => {
    if (listOPlayers.some((item) => item.name === textInputValue)) {
      return toast.error("You can't add same name", {
        position: toast.POSITION.TOP_RIGHT
      });
    }
    dispatch(addNewPlayerName(textInputValue));
    setTextInputValue("");
  }, [textInputValue]);

  const isAddButtonDisabled = textInputValue.length < 3;

  return (
    <Grid
      container
      direction="row"
      alignItems="center"
      spacing={1}
      marginBottom={5}
    >
      <Grid item>
        <TextField
          onChange={onInputHandleChange}
          size="small"
          label="New Player Name"
          onKeyPress={(event) => {
            if (event.key === "Enter") {
              event.preventDefault();
              if (!isAddButtonDisabled) {
                onAddButtonHandleClick();
              }
            }
          }}
          value={textInputValue}
        />
      </Grid>
      <Grid item>
        <Button
          onClick={onAddButtonHandleClick}
          disabled={isAddButtonDisabled}
          variant="contained"
        >
          Add New
        </Button>
      </Grid>
    </Grid>
  );
};

export default AddNewPlayerNameInput;
