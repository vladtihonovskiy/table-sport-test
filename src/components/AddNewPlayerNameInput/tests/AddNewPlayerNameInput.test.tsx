import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import * as reduxHooks from "react-redux";
import * as actions from "../../../store/slices/gamesStorySlice";
import AddNewPlayerNameInput from "../AddNewPlayerNameInput";

jest.mock("react-redux");

const mockedStoreSelector = jest.spyOn(reduxHooks, "useSelector");
const mockedDispatch = jest.spyOn(reduxHooks, "useDispatch");

describe("AddNewPlayerNameInput", () => {
  it("should render", () => {
    mockedStoreSelector.mockReturnValue({});

    const component = render(<AddNewPlayerNameInput />);

    expect(component).toMatchSnapshot();
  });

  it("should button be disabled by default", () => {
    mockedStoreSelector.mockReturnValue({});

    render(<AddNewPlayerNameInput />);

    expect(screen.getByRole("button")).toBeDisabled();
  });

  it("should button be disabled if input less then 3 symbols", () => {
    mockedStoreSelector.mockReturnValue({});

    render(<AddNewPlayerNameInput />);

    const input = screen.getByTestId("new-team-input");
    expect(input).toBeInTheDocument();

    fireEvent.change(input, { target: { value: "U" } });
    expect((input as HTMLInputElement).value).toBe("U");
    expect(screen.getByRole("button")).toBeDisabled();

    fireEvent.change(input, { target: { value: "Uk" } });
    expect((input as HTMLInputElement).value).toBe("Uk");
    expect(screen.getByRole("button")).toBeDisabled();

    fireEvent.change(input, { target: { value: "Ukr" } });
    expect((input as HTMLInputElement).value).toBe("Ukr");
    expect(screen.getByRole("button")).toBeEnabled();
  });

  it("should call addNewPlayerName action on button click and clear input", () => {
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedAddNewPlayerName = jest.spyOn(actions, "addNewPlayerName");
    mockedStoreSelector.mockReturnValue({ listOPlayers: [] });

    render(<AddNewPlayerNameInput />);

    const input = screen.getByTestId("new-team-input");
    const button = screen.getByRole("button");

    fireEvent.change(input, { target: { value: "Ukraine" } });
    fireEvent.click(button);
    expect(dispatch).toHaveBeenCalledTimes(1);

    expect(mockedAddNewPlayerName).toHaveBeenCalledWith("Ukraine");
    expect((input as HTMLInputElement).value).toBe("");
  });

  it("should not call addNewPlayerName with the same name", () => {
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedAddNewPlayerName = jest.spyOn(actions, "addNewPlayerName");
    mockedStoreSelector.mockReturnValue({
      listOPlayers: [{ name: "Ukraine" }],
    });

    render(<AddNewPlayerNameInput />);

    const input = screen.getByTestId("new-team-input");
    const button = screen.getByRole("button");

    fireEvent.change(input, { target: { value: "Ukraine" } });
    fireEvent.click(button);
    expect(mockedAddNewPlayerName).toHaveBeenCalledTimes(0);

    expect((input as HTMLInputElement).value).toBe("Ukraine");
  });
});
