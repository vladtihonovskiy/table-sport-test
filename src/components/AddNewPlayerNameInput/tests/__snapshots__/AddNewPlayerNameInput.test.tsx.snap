// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewPlayerNameInput should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-[object Object] css-b1jsqe-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-l7ms5s-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-109jk2q-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              New Team Name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="new-team-input"
                id=":r0:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1ftyaf0"
                >
                  <span>
                    New Team Name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-l7ms5s-MuiGrid-root"
        >
          <button
            class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-bbo1zh-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Add New
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-[object Object] css-b1jsqe-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-l7ms5s-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-109jk2q-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            New Team Name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="new-team-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  New Team Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-l7ms5s-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-bbo1zh-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Add New
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
